feat(contextswitcher): Change from dropdown to Bootstrap-select

rhamilto opened the pull request:
Fixes #1006

The contextswitcher was originally written using a dropdown, which doesn't provide the correct
functionality.  This PR preserves the existing styles for the dropdown for backward compatibility.

## Changes

Write a list of changes the PR introduces

* Replace dropdown with Bootstrap-select

## Link to rawgit and/or image

https://rawgit.com/rhamilto/patternfly/contextselector-selectpicker-dist/dist/tests/context-selector.html
https://rawgit.com/rhamilto/patternfly/contextselector-selectpicker-dist/dist/tests/context-selector-vertical-nav.html

## PR checklist (if relevant)
Note:  I am not currently able to cross-browser test IE as I don't have full access to BrowserStack at the moment.

- [x] **Cross browser**: works in IE9
- [x] **Cross browser**: works in IE10
- [x] **Cross browser**: works in IE11
- [x] **Cross browser**: works in Edge
- [x] **Cross browser**: works in Chrome
- [x] **Cross browser**: works in Firefox
- [x] **Cross browser**: works in Safari
- [x] **Cross browser**: works in Opera
- [X] **Responsive**: works in extra small, small, medium and large view ports.
- [X] **Preview the PR**: An image or a URL for designer to preview this PR is provided.

jeff-phillips-18 commented:
@rhamilto Can you run the regression tests for this? I assume there will be some updates.

rhamilto commented:
> @rhamilto Can you run the regression tests for this? I assume there will be some updates.

Ran against my branch and got 21 failures, none of which I caused (all appear to be related to missing primary nav items from horizontal nav).  Ran against master, got the same result.

rhamilto commented:
Per @dgutride, the test failures are a known issue and will be fixed with https://github.com/patternfly/patternfly/pull/991.

dabeng commented:
![ipeiahgfhl](https://user-images.githubusercontent.com/5689075/37692447-98f974e4-2cf3-11e8-87ff-e72748ac1fec.gif)

There are two highlighted items at the same time. My browser is Chrome Version 66.0.3359.33.

dabeng commented:
<img width="403" alt="screen shot 2018-03-21 at 10 41 05 am" src="https://user-images.githubusercontent.com/5689075/37692577-958520dc-2cf4-11e8-9d3b-5f2fe606cbd9.png">

Maybe this blank space is not expected.

matthewcarleton commented:
LGTM, other than what @dabeng pointed out, the only thing I saw was the padding on the  .dropown-toggle not using variables. Thanks for taking care of this @rhamilto !

rhamilto commented:
Thanks, @dabeng and @matthewcarleton.

> There are two highlighted items at the same time.

This is the default functionality of Bootstrap-select.  I'm not sure we'd want to change that for this specific context?

![bwnyc0wfj5](https://user-images.githubusercontent.com/895728/37717998-1aec9db0-2cf8-11e8-8ff0-5cefe9fd4799.gif)

> Maybe this blank space is not expected.

Indeed it is not.  Fixed.

![screen shot 2018-03-21 at 11 08 20 am](https://user-images.githubusercontent.com/895728/37718050-32ab0054-2cf8-11e8-998d-7b5da6fb4f46.PNG)

> the padding on the .dropown-toggle not using variables. 

Modified to use vars.

matthewcarleton commented:
I agree with @rhamilto we should keep the default bootstrap select functionality. I wonder if we need to keep all the deprecated css that I wrote for the old one. It would be nice to just remove it, What am I missing, is there a good reason to keep it? @LHinson what do you think?

rhamilto commented:
> is there a good reason to keep it?

Semantic versioning says you should keep it since removing it would break anyone using the dropdown version that already shipped.

dgutride commented:
@rhamilto @matthewcarleton  - I'd agree that we should keep it to prevent releasing a breaking change

matthewcarleton commented:
ok that makes sense. I wasn't sure if it was being used at all yet, I suppose that doesn't matter though. Thanks for the clarification @rhamilto @dgutride

jeff-phillips-18 commented:
@rhamilto Is this still WIP? Just due to cross-browser testing?

rhamilto commented:
> @rhamilto Is this still WIP? Just due to cross-browser testing?

Yes and yes.  Y'all have a BrowserStack account I can make use of?  @dtaylor113?

tlwu2013 commented:
Hey @rhamilto , I was just chatting with Serena and Colleen to see if we can remove the vertical separator of the context switcher? 

We ran into a tricky alignment issue between the vertical separator and the sidebar:
<img width="339" alt="tricky-misalignment-case" src="https://user-images.githubusercontent.com/5903705/38058890-52775fd0-3299-11e8-999f-694ccdc59556.png">

It's tricky because the logo length could be different, so if we can take out the left border of the switcher could help.  Thanks!

serenamarie125 commented:
> We ran into a tricky alignment issue between the vertical separator and the sidebar:

@tlwu2013 agreed, the vertical separator will cause an issue for both the cluster & app consoles.  I'm not sure that we should make this fix as part of this PR, because it may hold it up.

I've created a separate issue to track this : #1006

rhamilto commented:
Like this?

![screen shot 2018-03-29 at 8 25 32 am](https://user-images.githubusercontent.com/895728/38088748-e85d14cc-332a-11e8-86ec-e6b6cbce7391.PNG)
![screen shot 2018-03-29 at 8 25 52 am](https://user-images.githubusercontent.com/895728/38088749-e86799d8-332a-11e8-861f-1092bf6a0a93.PNG)

Note the utility nav items do not sit on the same baseline as the logo, so now the context pic

rhamilto commented:
(

rhamilto commented:
> @tlwu2013 agreed, the vertical separator will cause an issue for both the cluster & app consoles. I'm not sure that we should make this fix as part of this PR, because it may hold it up.

I think it's better to get this the way we want it before merging so there is one less deprecated version of the context switcher in PatternFly.

rhamilto commented:
PR updated with requested changes.  Once https://travis-ci.org/patternfly/patternfly/builds/359839029?utm_source=github_status&utm_medium=notification finishes, the test page will be updated.

beanh66 commented:
Thanks @rhamilto, that looks great!

jeff-phillips-18 commented on src/sass/converted/patternfly/_vertical-nav.scss:
```diff
@@ -129,6 +129,8 @@
       max-width: 120px;
       // If flexbox is supported, do not set max-width, take all space with just some right padding
+      // This generates a known issue on IE11:
+      // https://github.com/patternfly/patternfly/pull/810
```
this change can be removed as it was done in #1008

rhamilto commented:
Cross-browser testing complete.  Ready for final review.  [Build is failing](https://travis-ci.org/patternfly/patternfly/builds/361163857?utm_source=github_status&utm_medium=notification), but it's not clear as to why.

matthewcarleton commented:
@rhamilto this is looking great. I have two minor things to point out. 
1. As per the design doc there is meant to be an scroll on the list once it reaches a certain number of items (or a fixed height).
2. Do we want the longer strings to be truncated rather than breaking to a second line? Is that default for Bootstrap-select? 
@mcarrano what are your thoughts on these two items?

rhamilto commented:
> As per the design doc there is meant to be an scroll on the list once it reaches a certain number of items (or a fixed height).

That's possible, but would be an override to existing Bootstrap-select functionality. 

> Do we want the longer strings to be truncated rather than breaking to a second line? Is that default for Bootstrap-select?

That is also the default Bootstrap-select behavior, but I did add `text-overflow: ellipsis` so the text is ellipsized rather than just lopped off.

matthewcarleton commented:
@rhamilto I talked to @mcarrano about the truncated text and he is good with it so that can stay :) 

For the scrolling list I am going by the design doc so if we keep it this way we should update the doc. I'll leave that up to @mcarrano though.

jeff-phillips-18 commented:
@rhamilto Can you rebase please? I'd like to see if #1012 helps determine the build failure.

rhamilto commented:
Rebased.

jeff-phillips-18 commented:
@matthewcarleton I'd say approve this and then file an issue against either PF design or PF core to update whichever way we go. I think its ok to iterate on this as any changes in order to limit and scroll would not be breaking.

mcarrano commented:
@matthewcarleton So just to make sure that I understand the scrolling behavior, are we saying that the list would never produce a scroll bar? So what if there were 100 items to select from? Would the list just grow indefinitely? I don't think we would want that. Can file a separate issue, if necessary.

rhamilto commented:
> @matthewcarleton So just to make sure that I understand the scrolling behavior, are we saying that the list would never produce a scroll bar? So what if there were 100 items to select from? Would the list just grow indefinitely? I don't think we would want that. Can file a separate issue, if necessary.

Bootstrap-select will only make the dropdown as big as the available space (the dropdown stays within the viewport and a scrollbar is added to the dropdown).

![screen shot 2018-04-03 at 3 34 11 pm](https://user-images.githubusercontent.com/895728/38271499-a6c562e8-3754-11e8-986d-5f2904bd6b8f.PNG)

matthewcarleton commented:
@mcarrano @rhamilto I think this is fine. We should update the design doc to reflect this though.
